sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -437,12 +437,24 @@ pi_result _pi_context::initialize() {
   return PI_SUCCESS;
 }
 
-void _pi_context::finalize() {
+pi_result _pi_context::finalize() {
   // This function is called when pi_context is deallocated, piContextRelase.
   // There could be some memory that may have not been deallocated.
   // For example, zeEventPool could be still alive.
+  std::lock_guard<std::mutex> NumEventsLiveInEventPoolGuard(
+      NumEventsLiveInEventPoolMutex, std::adopt_lock);
   if (ZeEventPool && NumEventsLiveInEventPool[ZeEventPool])
     zeEventPoolDestroy(ZeEventPool);
+
+  // Destroy the command list used for initializations
+  ZE_CALL(zeCommandListDestroy(ZeCommandListInit));
+
+  // Destruction of some members of pi_context uses L0 context
+  // and therefore it must be valid at that point.
+  // Technically it should be placed to the destructor of pi_context
+  // but this makes API error handling more complex.
+  ZE_CALL(zeContextDestroy(ZeContext));
+  return PI_SUCCESS;
```

Comment:
that's good that you return status here now, but please check it in the caller

Reply:
Done.